tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,536 @@
+  Therefore, we can add them as placeholder inputs. Nodes other than
+  the remote ops and the placeholder ops can be considered as 'regular nodes',
+  where we can add them directly to the subgraph layer.
+
+  Args:
+    graph_def: A GraphDef proto for the original graph.
+    graph: A tf.Graph instance for the original graph.
+    node_name_to_node_def: A mapping from node names to 'NodeDef' protos.
+    previous_layers_visited: A set of node names that have been visited
+                             by the previously processed layers.
+    output_node_names: A set of output node names for the current layer.
+
+  Returns:
+    An ExecutionSpec representing a subgraph layer.
+  """
+  subgraph = graph_pb2.GraphDef()
+  subgraph.versions.CopyFrom(graph_def.versions)
+  subgraph.library.CopyFrom(graph_def.library)
+
+  queue = list(output_node_names)
+
+  # Nodes we've visited that belong to the current layer.
+  current_layer_visited = set([])
+  # Nodes coming from other layers.
+  nodes_from_other_layers = set([])
+
+  while queue:
+    current_node_name = queue[0]
+    current_node = node_name_to_node_def[current_node_name]
+    del queue[0]
+
+    if _is_remote_op(current_node) or _is_placeholder_op(current_node):
+      # These ops must be computed before this subgraph layer. Hence,
+      # we treat them as placeholder inputs.
+      if current_node_name not in current_layer_visited:
+        placeholder_node = _create_placeholder_node_from_existing_node(
+            current_node, graph)
+        subgraph.node.append(placeholder_node)
+
+        current_layer_visited.add(current_node_name)
+    else:
+      # A regular node coming from another subgraph layer may not be an
+      # output of that subgraph layer. We need to check for that later.
+      if current_node_name in previous_layers_visited:
+        nodes_from_other_layers.add(current_node_name)
+
+      elif current_node_name not in current_layer_visited:
+        subgraph.node.append(current_node)
+
+        current_layer_visited.add(current_node_name)
+        queue.extend(list(node_name_to_node_def[current_node_name].input))
+
+  return ExecutionSpec(
+      subgraph=subgraph,
+      input_names=_get_input_names(subgraph),
+      output_names=set(output_node_names),
+      is_remote_op=False,
+      body_node_names=_get_body_node_names(subgraph),
+      nodes_from_other_layers=nodes_from_other_layers)
+
+
+def _get_input_names(subgraph: Any) -> Set[Text]:
+  input_names = {node.name for node in subgraph.node
+                 if _is_placeholder_op(node)}
+  return input_names
+
+
+def _get_body_node_names(subgraph: Any) -> Set[Text]:
+  body_node_names = {node.name for node in subgraph.node
+                     if not _is_placeholder_op(node)}
+  return body_node_names
+
+
+def _handle_nodes_from_other_layers(
+    current_spec: ExecutionSpec,
+    execution_specs: List[ExecutionSpec],
+    graph: Any,
+    node_name_to_node_def: Mapping[Text, Any]
+) -> ExecutionSpec:
```

Comment:
Semantics of this function is quite unusual and hence the type signature a bit deceptive. It mutates one or more execution spec within the list `execution_specs` in addition to returning an updated spec which the caller then has to append to the same list. Not a big deal for a private utility but there are some obvious ways to improve:

The fact that this "handles other layers" is IMO besides the point. You could replace this function with:

```
def _merge_with_execution_specs(
    execution_spec: ExecutionSpec,
    execution_specs: Sequence[ExecutionSpec],
    ...) -> None
```

which suggests `execution_spec` is being merged with `execution_specs` (which does both "handling other layers" and appending). Note that the function need not return anything.

Reply:
Discussed above.